proto/asn1-pdu/asn1_universal_types_to_der.cc

```diff
@@ -23,6 +23,25 @@
 
 namespace asn1_universal_types {
 
+void Encode(const Boolean& boolean, std::vector<uint8_t>& der) {
+  // Save the current size in |tag_len_pos| to place tag and length
+  // after the value is encoded.
+  const size_t tag_len_pos = der.size();
+
+  // The contents octets shall consist of a single octet (X.690 (2015), 8.2.1).
+  if (boolean.val()) {
+    // If the boolean value is TRUE the octet shall have any non-zero value, as
+    // a sender's option (X.690 (2015), 8.2.2).
+    der.push_back(0xFF);
+  } else {
+    // If the boolean value is FALSE the octet shall be zero (X.690
+    // (2015), 8.2.2).
+    der.push_back(0x00);
+  }
+
+  EncodeTagAndLength(kAsn1Boolean, der.size() - tag_len_pos, tag_len_pos, der);
```

Comment:
This is mostly as an optimization, since this is a fairly 'fixed' field (like EOC)

```suggestion
der.push_back(kAsn1Boolean);  // Tag
der.push_back(0x01);  // Length; always one byte.
if (boolean.val()) {
  der.push_back(0xFF);
} else {
  der.push_back(0x00);
}
```

This just avoids the need for the extra length adjustment, since this is always at the end. I suppose I'd missed that along the way; a lot of these primitive types we always know the length from the Message itself (e.g. Integer is always `std::min(1, integer.val.size());`, BitString is always `bit_string.val.size() + 1`, etc

Reply:
In the course of writing up the above comment, I realized there's another bug: `Encode(UTCTime` and `Encode(GeneralizedTime` are bugged: you updated the signature for `EncodeTimestamp` but forgot to switch to using the bool.

The net effect is that the bool is always true, which means `EncodeUTCTime` always encodes as a `GeneralizedTime`

Reply:
@sleevi Good eye with the Time bug. I just fixed that. I also encode the tag and length immediately as suggested when we know the length from the Message. I was leaning towards that originally but didn't immediately do it, since I wanted to follow the `EncodeTagLength` pattern in the hope that it would make it easier to read.